### PR TITLE
Update Generics terminology document

### PR DESCRIPTION
**Pull request by josh11b**
This reflects changes from a number of approved proposals:
- #920 : concrete statements about orphan and overlap in Carbon
- #2138 : "generic" -> "checked generic", "template" -> "template generic"
- #2188 : binding patterns are forbidden in type position
- #2360 : "type", "facet type", "facet". Note: I am not using the term "generic type" from #2360 since that meaning conflicts with the generally accepted meaning of "generic type" of a type with a compile-time parameter.
- #2760 / #2770 : internal/external impl -> extending impl
- #2964 : "symbolic constant" and "template constant"

### CODE REVIEW

File: docs/design/generics/terminology.md [hunk 1]
```diff
@@ -107,7 +122,7 @@ Expected difference between generics and templates:
   <tr>
    <td>name lookup resolved for definitions in isolation ("early")
    </td>
-   <td>some name lookup may require information from calls (name lookup may be "late")
+   <td>name lookup can use information from calls (name lookup may be "late")
```
Comment (geoffromer) [inline docs/design/generics/terminology.md]: Optional drive-by fix:

```suggestion
   <td>name lookup can use information from arguments (name lookup may be "late")
```

File: docs/design/generics/terminology.md [hunk 2]
```diff
@@ -250,21 +265,111 @@ fail.
...
+A _type_ is a value of type `type`. Conversely, `type` is the type of all types.
+
+Expressions in type position, for example a [binding type expression](#bindings)
+or the return type of a function, are implicitly cast to type `type`. This means
+that it is legal to put a value that is not a type where a type is expected, as
+long as it has an implicit conversion to `type` that may be performed at compile
+time.
+
+## Facet type
+
```
Comment (geoffromer) [inline docs/design/generics/terminology.md:327]: After this PR, a lot of uses of the term "type-of-type" will link here. I assume you're planning to update them in a separate PR in order to keep this one small, but I'd prefer to minimize even temporary inconsistency. As an alternative to fixing all those usages in this PR, how about something like this?

```suggestion

> **TODO:** Documents using the obsolete term "type-of-type" should be
> updated to say "facet type" instead.

```

File: docs/design/generics/terminology.md [hunk 3]
```diff
@@ -250,21 +265,111 @@ fail.
...
+`type`, so it is a facet that is not a type. However, in places where a type is
+expected, for example in a [binding type expression](#bindings) or after the
+`->` in a function declaration, there is an automatic implicit conversion to
+`type`. This means that a facet may be used in those positions. For example, the
+facet `i32 as Hashable` will implicitly convert to `(i32 as Hashable) as type`,
+which is `i32`, in those contexts.
+
+## Type expression
+
+A _type expression_ is an expression that is being used as a type. In some
```
Comment (geoffromer) [inline docs/design/generics/terminology.md]: I continue to be uncomfortable defining "type expression" to be a property of the context, not a property of the expression itself. Would this work instead?

```suggestion
A _type expression_ is an expression that can be used as a type. In some
```

File: docs/design/generics/terminology.md [hunk 4]
```diff
@@ -250,21 +265,111 @@ fail.
...
+cases, what is written in the source code is a value, like a [facet](#facet) or
+tuple of types, that is not a type but has an implicit conversion to `type`. In
+those cases, we are concerned with the type value after the implicit conversion.
+
+## Facet binding
+
+We use the term _facet binding_ to refer to the name introduced by a
+[constant binding pattern](#bindings) (using `:!` with or without the `template`
+modifier) with a [facet type](#facet-type). In the binding pattern
```
Comment (geoffromer) [inline docs/design/generics/terminology.md]:
```suggestion
modifier) where the declared type is a [facet type](#facet-type). In the binding pattern
```

File: docs/design/generics/terminology.md [hunk 5]
```diff
@@ -349,27 +452,24 @@ instead of associated entity.
...
+implement a named constraint as a way to implement all of the interfaces it
+requires.
 
-### Internal impl
+### Extending an impl
 
-A type that implements an interface _internally_ has all the named members of
-the interface as named members of the type. This means that the members of the
-interface are available by way of both
+A type that _extends_ the implementation an interface has all the named members
```
Comment (geoffromer) [inline docs/design/generics/terminology.md]: This terminology suggests that there's a point in time where the implementation exists, but hasn't yet been extended, which doesn't match my intuition (or the syntax, where `extend` comes first). Could we instead say the type "extends the interface", or "implements the interface as an extension"?

File: docs/design/generics/terminology.md [hunk 6]
```diff
@@ -349,27 +455,24 @@ instead of associated entity.
...
```
Comment (geoffromer) [inline docs/design/generics/terminology.md]:
```suggestion
A type that _extends_ the implementation of an interface has all the named members
```

File: docs/design/generics/terminology.md (hunk 5 above)
Comment (josh11b) [reply docs/design/generics/terminology.md]: The _syntax_ we decided on is `extend impl as`..., so I'm trying to make the terminology match. We are going to have other kinds of extending (base class, mixin), so whatever terminology we use should hopefully work consistently across those cases.

File: docs/design/generics/terminology.md [hunk 7]
```diff
@@ -250,21 +265,114 @@ fail.
...
+declare function parameters, in `let` and `var` declarations, as well as to
+declare [generic parameters](#generic-means-compile-time-parameterized). There
+are three kinds of binding patterns, corresponding to
+[the three value phases](/docs/design/README.md#value-categories-and-value-phases):
+
+-   A _runtime binding pattern_ binds to a dynamic value at runtime, and is
+    written using a `:`, as in `x: i32`.
+-   A _symbolic constant binding pattern_ or _symbolic binding pattern_ binds to
+    a compile-time value that is not known when type checking, and is used to
+    declare [checked generics](#checked-versus-template-parameters) parameters.
```
Comment (zygoloid) [inline docs/design/generics/terminology.md]:
```suggestion
    declare [checked generic](#checked-versus-template-parameters) parameters.
```

File: docs/design/generics/terminology.md [hunk 8]
```diff
@@ -736,40 +842,39 @@ types for which no implementation exists will be rejected at the call site:
 DoAdd(apple, orange);
 ```
 
+The type of an interface parameters and associated constants is commonly a
+[facet type](#facet-type), but not always. For example, one might have an
+integer type and be used to specify the size of an array type.
```
Comment (zygoloid) [inline docs/design/generics/terminology.md]: I misread "one might have an integer type" as "[a person] might have an integer type" at first. Maybe:

```suggestion
[facet type](#facet-type), but not always. For example, an interface parameter
that specifies an array bound might have an integer type.
```

**Review by zygoloid — APPROVED** [review]

File: docs/design/generics/terminology.md (hunk 5 above)
Comment (josh11b) [reply docs/design/generics/terminology.md]: @zygoloid pointed out to me that it is in fact possible (though weird) to implement an interface for a type that is forward declared, and it only extends when the class is later defined.

Comment (zygoloid) [reply docs/design/generics/terminology.md]: I think you would be able to define the implementation first and extend it later:
```carbon
class A;
impl A as X { fn F() {} }
class A {
  extend impl A as X;
}
```
... though this is certainly not going to be the common case. But I think the implication that defining the implementation and extending it are separate steps is a correct implication.
